packages/embark-code-runner/src/index.ts

```diff
@@ -52,101 +50,57 @@ export default class CodeRunner {
     });
   }
 
-  private fireEmbarkJSEvents(code: string) {
-    const regexRegister = /EmbarkJS\.(.*)\.registerProvider/gm;
-    const regexSet = /EmbarkJS\.(.*)\.setProvider/gm;
-    let matches = regexRegister.exec(code);
-    if (matches) {
-      let [, provider] = matches;
-      provider = provider.toLowerCase();
-      this.providerStates[`${provider}:${ProviderEventType.ProviderRegistered}`] = true;
-      this.events.emit(`runcode:${provider}:${ProviderEventType.ProviderRegistered}`);
-    }
-    matches = regexSet.exec(code);
-    if (matches) {
-      let [, provider] = matches;
-      provider = provider.toLowerCase();
-      this.providerStates[`${provider}:${ProviderEventType.ProviderSet}`] = true;
-      this.events.emit(`runcode:${provider}:${ProviderEventType.ProviderSet}`);
-    }
-  }
-
   private registerEvents() {
-  this.events.on("runcode:register", this.registerVar.bind(this));
-
-  this.events.on("runcode:init-console-code:updated", (code: string, cb: Callback<null>) => {
-    this.evalCode(code, (err, _result) => {
-      if (err) {
-        this.logger.error("Error running init console code: ", err.message || err);
-      }
-      this.fireEmbarkJSEvents(code);
-      cb();
-    });
-  });
+    this.events.on("runcode:register", this.registerVar.bind(this));
+  }
 
-  this.events.on("runcode:embarkjs-code:updated", (code: string, cb: Callback<any>) => {
-    this.evalCode(code, (err, _result) => {
-      if (err) {
-        this.logger.error("Error running embarkjs code: ", err.message || err);
-      }
-      this.fireEmbarkJSEvents(code);
-      cb();
+  private registerCommands() {
+    this.events.setCommandHandler("runcode:getContext", (cb) => {
+      cb(this.vm.options.sandbox);
     });
-  });
-}
+    this.events.setCommandHandler("runcode:eval", this.evalCode.bind(this));
+    this.events.setCommandHandler("runcode:embarkjs:reset", this.resetEmbarkJS.bind(this));
+    this.events.on("contractsDeployed", this.resetEmbarkJS.bind(this));
+
+    // register listeners for when EmbarkJS runs registerProvider through the console.
+    // For example, when `EmbarkJS.Storage.registerProvider(...)` is run through the console,
+    // emit the `runcode:storage:providerRegistered` event, and fire any requests attached to it
+    Object.keys(EmbarkJS)
+      .filter((propName) => EmbarkJS[propName].hasOwnProperty("registerProvider"))
```

Comment:
It looks like `hasOwnProperty()` is not needed here (unless I'm missing something).

Usually this is used in combo with `for .. in ..` loops as they iterate over an object's prototype as well. In order to ensure only properties of the object, not its prototype, are processed we **used** to do:

```
for (let sth in sthelse) {
  if (sthelse.hasOwnProperty(sth) {
    // we're good to go
  }
}
```

Using the more modern `for .. of ..` syntax, you get this for free:

```
for (let sth of sthelse) {
  // no hasOwnPropertyCheck needed 
}
```

Having that said, in this case here, we're simply doing a property lookup on the `EmbarkJS` object and there's no `registerProvider` property/method on `EmbarkJS`' prototype. In other words, we might as well just do

```
...
.filter(propName => !!EmbarkJS[propName].registerProvider)
```

Hope this makes sense!

Reply:
I'm fairly certain this section is code that got moved not written fresh. That being said, the improvements you've suggested do make sense.

Reply:
Right. Yea maybe this is better addressed in a future PR then. I just feel that those kind of things just get forgotten (since they also don't have a super huge impact), so sometimes situations like these are good opportunities to improve the state of affairs.

No biggie either way. Happy to have it changed in another effort.